\documentclass[12pt]{article}
\usepackage[utf8]{inputenc}

\usepackage[margin=2cm]{geometry}
\usepackage{times}

\usepackage{lineno}

\usepackage{hyperref}
\hypersetup{
    colorlinks=true,
    linkcolor=blue,
    filecolor=magenta,      
    urlcolor=cyan,
}

\title{OvRF review}
\author{Laura Mu\~noz Baena and Art Poon}
\date{2020}

\begin{document}

\maketitle

\pagewiselinenumbers

\section {Introduction}
Overlapping genes are nucleotide sequences that encode more than one protein. 
There is a general consensus on the reasons to explain overlapping genes in viruses. 
The compression idea states that this type of gene arrangement allow viruses to maximize the amount of information contained in their genomes\cite{lamb1991diversity, pavesi1997informational}. 
This hypothesis is consistent with the idea that viral genomes tend to be small in order to prevent deleterious mutations and carry less replication times\cite{belshaw2008pacing}, improving their infectivity. However, the discovery of overlapping genes in longer genomes including both prokaryotic\cite{normark1983overlapping, johnson2004properties, palleja2008large} and eukaryotic organisms\cite{spencer1986overlapping, williams1986mouse, makalowska2005overlapping} challenges the compression theory.
A second theory suggests that overlapping genes could have potential regulatory roles\cite{normark1983overlapping, kapranov2007genome}.
This was first suggested by Normak and collaborators\cite{normark1983overlapping}, which examined overlapping genes in \textit{Escherichia coli}, and described the possible implications in regulation of gene expression. 
In viruses, the effects of overlapping genes in transcription and translation, have been studied for example in the Autographa californica nuclear polyhedrosis virus (AcNPV)\cite{friesen1985temporal}, the Bacteriohage MS2\cite{scherbakov2000overlapping} and Simian virus 40(SV40)\cite{buchman1984complex}.
The idea that overlapping regions could be used by viruses to create \textit{de novo} genes by a process known as "overprinting" was first suggested by Grasse\cite{grasse2013evolution} in 1973.
Even though several papers on this topic have been written since then, the core ideas regarding overprinting process remain the same: 
1. In eukaryotes, genes product of overprinting are duplicated and separated in different sections of the genome, but in viruses genes tend to remain overlapped due to size constrains. 
2. Genes that comes from overprinting have an unusual codon usage and encode proteins that have diverse functions specialized to the current life-style of the organism in which they are found. 
More recent studies have shown that overlapping regions tend to have higher levels of intrinsic structural disorder, likely because one of the members was recently born \textit{de novo}\cite{willis2018gene}.
The way overlapping genes arrange vary according to every organism. However, there are five frame shift possibilities for each overlap in reference with the parental gene (which we refer as +0 reading frame).
If the genes are in the positive strand, there can be genes with a frame shift of +1 or +2.
If the genes are in the negative strand, there can be genes in -0 (opposite strand without shifting), -1 or -2 reading frames.

\section{Materials and Methods}
\subsection{Overview of overlapping reading frames}
Initially, we downloaded an accession list of all viral genomes from NCBI \url{https://www.ncbi.nlm.nih.gov/genome/viruses/} that has information regarding: representative, neighbor, host, taxonomy and segment name for every virus entry on the database. 
% Re-making this part of the analysis. Will have to updated scripts and methods
We then used a Python script to complement information of the entries to include taxonomical information, genome lengt and total number of proteins. 
The same script was used to download all of the proteins associated with each accession number and their position on the genome. 
We used the protein file to find overlapping proteins and calculate the frameshift and number of nucleotides involved in each overlap/ 
We also created a script to classify each viral family according to the Baltimore classification displayed on ViralZone \url{https://viralzone.expasy.org/}.
The final result is a data base with a total of 11,891 viral entries, from which 9,928 are classified in any Baltimore groups. 
From remaining 1,963 entries, we can say, according to the information obtained from NCBI, that 400 are DNA viruses and 1,552 are RNA. 


\subsection{Adjacency graphs}
To analyze the organization and distribution of overlapping reading frames in different virus families, we downloaded for each family all the reference complete genomes from \url{https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&SourceDB_s=RefSeq} as a fasta file. 
Over that file, we run a Python script to measure kmer-distance to create a distance matrix.
We then used that matrix as an input to form clusters representing similar proteins. 
Once we have each protein classified in a specific cluster, we generate a network plot, where each node represents a cluster, and the edges between clusters are formed when two proteins are adjacent to each other. 



\newpage
\bibliography{main}
\bibliographystyle{vancouver}

\end{document}
